grpc-contrib/src/main/java/com/salesforce/grpc/contrib/Statuses.java

```diff
@@ -0,0 +1,111 @@
+/*
+ *  Copyright (c) 2017, salesforce.com, inc.
+ *  All rights reserved.
+ *  Licensed under the BSD 3-Clause license.
+ *  For full license text, see LICENSE.txt file in the repo root  or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+package com.salesforce.grpc.contrib;
+
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.StatusException;
+import io.grpc.StatusRuntimeException;
+
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+
+/**
+ * Statuses is a collection of utility methods for working with gRPC {@code Status} responses.
+ *
+ * <p>gRPC statuses are annoying to work with because they tend to come bundled in either a {@code StatusException} or a
+ * {@code StatusRuntimeException}, but {@code StatusException} and {@code StatusRuntimeException} are unrelated classes,
+ * so working with them requires frequent {@code instanceof} checks and downcasting. The operations in this class try
+ * to smooth over these minor annoyances.
+ *
+ * <p>For example:
+ * <pre>
+ *     import static com.salesforce.grpc.contrib.Statuses;
+ *
+ *     try {
+ *         stub.someServiceOperation(...);
+ *     } catch (Throwable t) {
+ *         if (hasStatusCode(t, Status.Code.UNAUTHENTICATED) {
+ *             doWithStatus(t, (status, metadata) -> showLoginPrompt();
+ *         } else if (Statuses.hasStatus(t) {
+ *             doWithStatus(t, (status, metadata) -> handleGrpcProblem(status);
+ *         } else {
+ *             throw t;
+ *         }
```

Comment:
maybe empower doWithStatus with a return boolean, to condense out the if's, roughly like:

catch (Throwable t) {
    boolean handled = doWithStatus(t, Status.Code.UNAUTHENTICATED, (status, metadata) -> showLoginPrompt()) ||
        doWithStatus(t, (status, metadata) -> handleGrpcProblem(status));
    
    if (!handled) throw t;

}

Reply:
Hmm...I kind of like the use of flow control statements for flow control. There is a variant of `doWithStatus()` that takes a `Function<T>`, so you could accomplish something similar with the existing API.